horqrux/shots.py

```diff
@@ -0,0 +1,41 @@
+from __future__ import annotations
+
+import jax
+import jax.numpy as jnp
+from jax import Array
+
+from horqrux.apply import apply_gate
+from horqrux.primitive import Primitive
+
+
+def finite_shots(
+    state: Array,
+    gates: list[Primitive],
+    observable: Primitive,
+    values: dict[str, float],
+    n_shots: int = 100,
+) -> Array:
+    """
+    Run 'state' through a sequence of 'gates' given parameters 'values'
+    and compute the expectation given an observable.
+    """
+    state = apply_gate(state, gates, values)
+    # NOTE this only works now for an observable comprised of a single gate
+    # to get eigvals,eigvecs for arbitary compositions of paulis, we need to
+    # create the full tensor. check `block_to_jax` in qadence for this
+    eigvals, eigvecs = jnp.linalg.eig(observable.unitary())
```

Comment:
I think it should be the generator here instead of the unitary.

And since the generator is hermitian, we should be able to use `linalg.eigh` to get the eigenvalues/vectors more efficiently.

But I can tidy this stuff.

Reply:
Ah, I was under the impression that observables were arbitrary Hamiltonians. I got this bit wrong: it turns out that we are currently focussing on observables that are both hermitian and unitary, so this looks fine after all.